File: Course-Schedule.java
1class Solution {
2    List<List<Integer>> adj_list;
3    int[] indegree;
4    public boolean canFinish(int numCourses, int[][] prerequisites) 
5    {
6        indegree = new int[numCourses];
7        adj_list = new ArrayList<List<Integer>>();
8        for(int i=0;i<numCourses ;i++)
9        {
10            List<Integer> neighbor= new ArrayList<Integer>();
11            adj_list.add(neighbor);
12        }
13        for(int i=0;i<prerequisites.length;i++)
14        {
15            adj_list.get(prerequisites[i][1]).add(prerequisites[i][0]);
16            indegree[prerequisites[i][0]]++;
17        }
18        List<Integer> currList = new ArrayList<Integer>();
19        for(int i=0;i<numCourses;i++)
20        {
21            if(indegree[i] == 0)
22                currList.add(i);
23        }
24        int poppedCount= 0;
25        while(!currList.isEmpty())
26        {
27            List<Integer> child = new ArrayList<Integer>();
28            while(!currList.isEmpty())
29            {
30                int course = currList.removeFirst();
31                poppedCount++;
32                List<Integer> neighbor = adj_list.get(course);
33                for(int i=0;i< neighbor.size();i++)
34                {
35                    indegree[neighbor.get(i)]--;
36                    if(indegree[neighbor.get(i)] == 0)
37                        child.add(neighbor.get(i));
38                }
39            }
40
41            currList.addAll(child);
42        }
43
44    if(poppedCount == numCourses)
45            return true;
46    return false;   
47    }
48    
49}

File: Course-Schedule.cpp
1class Solution {
2public:
3    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) 
4    {
5        vector<vector<int>> adjList;
6        vector<int> inDegree;    
7        queue<int> currList;
8        for(int i=0;i<numCourses;i++)
9        {
10            inDegree.push_back(0);
11            vector<int> list;
12            adjList.push_back(list);
13        } 
14
15        for(int i=0;i<prerequisites.size();i++)
16        {
17            int dest = prerequisites[i][0], source = prerequisites[i][1];
18            adjList[source].push_back(dest);
19            inDegree[dest]++;
20        }
21        
22        for(int i=0;i<numCourses;i++)
23        {
24            if(inDegree[i] == 0)
25                currList.push(i);
26        } 
27        vector<int> order;
28        while(!currList.empty())
29        {
30            int course = currList.front();
31            currList.pop();
32            order.push_back(course);
33            vector<int> dependents = adjList[course];
34            for(int dependent: dependents )
35            {
36                inDegree[dependent]--;
37                if(inDegree[dependent] == 0)
38                    currList.push(dependent);
39            }
40        }
41
42        if(order.size() == numCourses)
43            return true;
44        return false;
45
46    }
47};